### PR TITLE
add cosign_copy

### PR DESCRIPTION
**Pull request by imjasonh**
I haven't tested this locally since it requires ambient creds to sign, but I've at least tested that the `cosign_verify` fails with the expected warning about missing ambient creds.

### CODE REVIEW

File: docs/resources/copy.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `destination_repo` (String) The destination repository.
+- `source` (String) The digest of the container image to copy.
```
Comment (mattmoor) [inline docs/resources/copy.md:21]: Hmm, should it take a tag?  If the destination is a repo, then what tag is applied?

File: internal/provider/resource_copy_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+		Steps: []resource.TestStep{{
+			// Sign and copy the image, then verify the copy's signature.
+			Config: fmt.Sprintf(`
+resource "cosign_sign" "foo" {
+  image = %q
+}
```
Comment (mattmoor) [inline internal/provider/resource_copy_test.go:47]: You should be able to `cosign_attest` and check that is copied too!

File: internal/provider/resource_copy.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			"source": schema.StringAttribute{
+				MarkdownDescription: "The digest of the container image to copy.",
+				Optional:            false,
+				Required:            true,
+				Validators:          []validator.String{validators.DigestValidator{}},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"destination_repo": schema.StringAttribute{
```
Comment (mattmoor) [inline internal/provider/resource_copy.go:66]: I don't love how verbose this is.  Can we make it `destination` and let the type qualify it?

File: internal/provider/resource_copy.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	ropts := options.RegistryOptions{
+		KubernetesKeychain: true,
+	}
+	dst, err := name.NewRepository(data.DestinationRepo.ValueString())
+	if err != nil {
+		return "", errors.New("Unable to parse destination repository")
+	}
+
+	if err := copy.CopyCmd(ctx, ropts, digest.String(), dst.String(), false, false); err != nil {
```
Comment (mattmoor) [inline internal/provider/resource_copy.go:100]: I want to confirm that the semantics of this are to create the input reference in the `dst` repo, so if a tag is passed, the same tag is copied, but if a digest is passed the digest is copied (without tagging).

My concern here seeing the `dst.String()` is that the repo will be interpreted as having an implicit `:latest` and that we cannot use this without always overwriting `:latest`.

Comment (mattmoor) [reply internal/provider/resource_copy.go:100]: If we made the target `dst.Digest(digest.DigestStr()).String()` that would be maximally clear, but not sure if `cosign` accepts that 😅

File: docs/resources/copy.md (hunk 1 above)
Comment (imjasonh) [reply docs/resources/copy.md:21]: My thinking was that a separate `oci_tag` would be responsible for tagging. `cosign` shouldn't think about tags at all, IMO.

I don't feel that strongly though, and supporting tags would match `cosign` CLI behavior.

File: internal/provider/resource_copy.go (hunk 4 above)
Comment (imjasonh) [reply internal/provider/resource_copy.go]: Let's try it. I agree that `CopyCmd`'s args don't make it clear whether it takes a repo, a repo(-but-then-also-apply-`:latest`), or would accept a digest.

File: internal/provider/resource_copy_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+resource "cosign_attest" "foo" {
+  image          = %q
```
Comment (mattmoor) [inline internal/provider/resource_copy_test.go:50]: make this signed_ref

File: internal/provider/resource_copy_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  predicate_type = "https://predicate.type"
+  predicate      = jsonencode({
+    foo = "bar"
+  })
+}
+
+resource "cosign_copy" "copy" {
+  source      = cosign_sign.foo.signed_ref
```
Comment (mattmoor) [inline internal/provider/resource_copy_test.go:58]: make this attested_ref

**Review by mattmoor — COMMENTED** [review]
You created a race in the test

**Review by mattmoor — COMMENTED** [review]
You created a race in the test

File: internal/provider/resource_copy_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+resource "cosign_attest" "foo" {
+  image          = cosign_sign.foo.signed_ref
+  predicate_type = "https://predicate.type"
+  predicate      = jsonencode({
+    foo = "bar"
+  })
+}
+
+resource "cosign_copy" "copy" {
+  source      = cosign_sign.foo.attested_ref
```
Comment (mattmoor) [inline internal/provider/resource_copy_test.go:58]:
```suggestion
  source      = cosign_attest.foo.attested_ref
```

**Review by mattmoor — APPROVED** [review]

**Pull request merged**